pip/src/fs.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use std::{
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+
+use miette::miette;
+use pyo3::{
+    exceptions::PyException,
+    prelude::*,
+    types::{PyDict, PyList, PyString, PyTuple},
+};
+use qsc::project::{DirEntry, EntryType, FileSystem};
+
+pub(crate) fn file_system(
+    py: Python,
+    read_file: PyObject,
+    list_directory: PyObject,
+) -> impl FileSystem + '_ {
+    Py {
+        py,
+        fs_hooks: FsHooks {
+            read_file,
+            list_directory,
+        },
+    }
+}
+
+struct FsHooks {
+    read_file: PyObject,
+    list_directory: PyObject,
+}
+
+#[derive(Debug)]
+struct Entry {
+    entry_type: EntryType,
+    path: String,
+    entry_name: String,
+    extension: String,
+}
+
+impl DirEntry for Entry {
+    type Error = pyo3::PyErr;
```

Comment:
I had to pick an error type because `entry_type` on the trait returns a `Result` . It's true none of the implementations can actually fail. What would be an error type to pick in this case?